### ISSUE TITLE
feature request: handle non perl dependencies

### ISSUE THREAD

**Issue by happy-barney** [Perl-Toolchain-Gang/CPAN-Meta#79]
Provide unified way how to handle non perl dependencies, eg C libraries.

Proposal:
extend current:
  requires => { package => version }

to:
  requires => { 'ACME::Extra::Dependencies::PkgConfig' => [
    { mod => 'glib-2.0', min_version => '2.42.0' },
    { mod => 'gtk+3.0' },
  ]}

**Comment by jberger** [comment]
> Here's an alternative that doesn't depend on "x_" type keys (which I think goes against the benefits of having a standard way of specifying third-party and system dependencies), and doesn't potentially create a forest of custom solutions like the Alien:: namespace is doing.

I honestly don't see how that solution would meet any of those criteria. Adding dependencies in an incompatible manner would break existing clients which wouldn't know what to do with them, it is the reason you'd want a new key which you'd prototype with x_. And it would necessarily involve a forest of solutions to handle the different uri types, unless it was just supposed to be advisory in which case you're breaking lots of things for little gain. Finally, Alien::Build/Base has a reason that it is the way it is, it will never make global changes to the system, angering sysadmins. It will either use a pre-installed dep or it will build/fetch a copy to be used locally.  It took lots and lots of design to make that work, especially given all the various platforms and build systems.

**Comment by Leont** [comment]
> Don't the current clients out there check if they support meta_spec version and then complain or prompt for upgrading if not?

Complain and fail to install is not generally considered an acceptable outcome. Having a meta 3.x next to meta 2.0 might be acceptable, but for a third file doing almost exactly the same to be worth it, it would have to be pretty damn solid (instead of all the uncertainties I read above)

**Comment by sjn** [comment]
[URI::PackageURL](https://metacpan.org/pod/URI::PackageURL) might offer an avenue for this feature, no?

**Comment by jberger** [comment]
It is interesting to see that someone is considering a universal mechanism of addressability for dependencies like this. However, given that we're talking about system-level dependencies there would still be the problem of not having a single platform/architecture/etc. You'd need to specify likely many different alternatives to install from debian/rpm/windows/mac sources, and even then you'd certainly never get them all.

I think it still would be best to have some standard key, perhaps added in a later 2.x if it must, or `x_` prefixed of course, that indicates that there are system dependencies and if possible what they are, perhaps using something like this new format. However, it would still be better if your Makefile.PL (or the like) then knows how to detect its presence and indicate failure. Installation tools that point to system packages will likely never be able to do that.

That said, once again, that was literally the design purpose of Alien::Base. As extended by @plicease you write an alienfile that knows how to detect if a system dependency already exists. The consuming module can then choose if it should be installed if not already present or if it should just bail out. (Also note that A::B never installs system packages because it is not within the principle of least surprise that installing a perl module, possibly locally, will install a system dep, and if it did systems administrators would kill us).

**Comment by plicease** [comment]
> That said, once again, that was literally the design purpose of Alien::Base. As extended by @plicease you write an alienfile that knows how to detect if a system dependency already exists. The consuming module can then choose if it should be installed if not already present or if it should just bail out. (Also note that A::B never installs system packages because it is not within the principle of least surprise that installing a perl module, possibly locally, will install a system dep, and if it did systems administrators would kill us).

I agree with all of this.  I personally do not think a cpan client (or Perl modules) should ever by default install system packages, but having an opt-in or plugin system may be appropriate.  IMO it probably won't ever be able to support every platform, but supporting the three or four that cover 99% of systems that are used in practice would probably be "good enough" especially if it is designed in a way that it doesn't exclude adding other platforms later.

I can think of a couple of useful ways of handling this.

1. meta provides the appropriate values for discovering the package via `Devel::CheckLib`.  This is really only enough to print an error message that says "you need to install xyz library".  It may also not work well on Windows for some subset of Windows Perls or some subset of libraries on Windows (where libraries sometimes have a slightly different name; this can also happen on non-Windows).

2. in addition to the meta that does detection in (1), include the Alien that will provide the library dynamically.  Just specifying the Alien as a dependency will functionally do the same thing in the normal case, but if the cpan client can do the library detection it may save some dependencies in the case of a system install (ie the system already provides the needed library).  One downside is that older clients that do not recognize the new meta will have no way of knowing to install the Alien.

3. in addition to the meta that does detection in (1) either provide the native library name and version, or compute the native library name and version _somehow_.  The Spec could also provide for specifying a fallback Alien as described in (2).  This way the preferences would be a) try the system lib using Devel::CheckLib b) install the system package **if the user has opted into that functionality**  c) fall back on the specified Alien if that was provided.

4. guess you could also do everything in (3) except the Alien fallback; I personally am against that since I think the Alien concept is useful.

I don't think the first option adds much.

The second I think has some merit in that it does provide some utility with a smaller amount of effort.  The second option is more or less what I do with FFI modules now (except substitute FFI::CheckLib and it is handled in the module installer logic instead of the cpan client), and it works pretty well.

The third option I think offers the most flexibility for cpan authors and users of those packages, but it also adds extra complexity.

Most options duplicates at least some of the work that Alien is already doing.  This is just my 2¢.  FWIW I've been thinking about these options for a while, but I don't think I have ever written them down anywhere.

**Comment by sjn** [comment]
Good perspectives!

For me, I think it's worth reminding ourselves that CPAN::Meta's purpose is to communicate that different types of dependencies are required (or recommended) for the the affected software to function as expected, and that the question in this ticket is how we can do it in a manner that allows us to get the necessary information to resolve these dependencies (or if that's not possible, as you say, to offer some instructions).

Package URLs are seeming to become a standard way for exactly this, especially in the SBOM space, where security professionals are required by law to ensure that they have a complete picture of where their software components come from, and how/who have modified it on the way from the author/vendor to the user.

For us, this could look as simple as this:

Before:
```
requires => { package => version },
```
New:
```
requires => { package => version },
external => { comment => [ pkgurl_alt1, pkgurl_alt2 ] }
```

…where pkgurl_altN is a valid PackageURL (as in the [spec](https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst)) possibly implemented by [URI::PackageURL](https://github.com/giterlizzi/perl-URI-PackageURL)?

I guess PackageURLs without a version should be interpreted as "use the latest", but AFAICT the PURL spec doesn't have any examples of how to specify version ranges, etc, so it seems they are mainly intended for reporting what has been installed (e.g. as part of an SBOM).

I still think PURLs could be used for specifying requirements (with a version range) if we make a careful proposal to the spec and the URI::PackageURL author? :grin:.

Then there's the generic PURL type: https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#generic :smile:
